The OpenMC C API

paulromano opened the pull request:
This pull request introduces a new C API for OpenMC that can be used for direct, in-memory coupling rather than having to repeatedly run the `openmc` executable (as OpenDeplete does currently). I want to stress that this is a work-in-progress and the API is likely to change/expand over time, but I wanted to at least put the initial capability out here in case others want to start using it.

To define C functions, I've used Fortran/C interoperability as defined by the F2008 standard. Namely, to have a procedure in Fortran be interoperable with (and callable from) C, one needs to add `bind(C)` and make sure that any arguments have interoperable type parameters. The Fortran standard provides a `ISO_C_BINDING` intrinsic module that defines kind type parameters for intrinsic types that correspond directly to C types. As one example, `integer(kind=C_LONG_LONG)` would correspond to `long long` in C. By adding `bind(C)` to a Fortran procedure, it ensures that the compiler will not perform any name mangling and the procedure will be available in the symbol table with whatever name you gave it.

A new module in the Fortran code called `openmc_api` makes all the C-interoperable procedures available (thus, if you wanted to write a Fortran wrapper that makes use of this API, you could just `use openmc_api`). Three of these procedures already exist: `openmc_init`, `openmc_run`, and `openmc_finalize` and are simply used by the openmc_api module. However, the module also defines a bunch of new procedures that allow you to modify the temperature/densities of cells/materials, reset tallies, load nuclides from the HDF5 library, etc.

On the Python side, I've created bindings to these "C functions" by using the ctypes standard-library module, which allows one to call functions from a shared library. In `openmc/capi.py`, there is a new `OpenMCLibrary` class for which a single instance (`openmc.lib`) is created at import-time. This class has methods, each of which corresponds to one of the functions from the C API. So, the following would be equivalent to running the `openmc` executable:
```Python
import openmc

openmc.lib.init()
openmc.lib.run()
openmc.lib.finalize()
```
The methods of `OpenMCLibrary` handle conversion of Python types to C types where necessary. For example, `openmc_set_temperature` expects an array of doubles to be passed, so `OpenMCLibrary.set_temperature` will convert a tuple/list/numpy array to a normal C array of doubles.

I had to make a number of changes to get things to play nicely:

- To run multiple simulations, one should call `openmc_reset` in between iterations. Additionally, to not re-use the same random numbers, I had to introduce a new global variable `total_gen` that keeps track of the overall number of generations
- When tallies.out is written, before we would actually change the tally results array (turning it into means / standard deviations). I've changed it so that the results array is not changed in-place and instead a `mean_stdev` function gets the mean and standard deviation of each bin.
- For coupling purposes, I've introduced a way for the user to specify a range of temperatures for which cross section should be loaded at the beginning of a simulation so that if the temperature of materials changes, new cross sections don't have to be loaded later.

For whoever is reviewing this PR, if you're confused about the motivation for any of my changes, just ask and I'll try to explain.

The whole thing works really nice. The only part that is a little hacky is how the Python module loads the shared library -- I had to have CMakeLists.txt copy the shared library from the build directory to the Python module directory. If anyone has a better idea for how to do that, I'm all ears.

I'm planning on refactoring OpenDeplete to use the C API so that there are no "write to disk/read from disk" steps at each iteration. Instead, all data can be transferred in memory. We're also planning on leveraging the C API for thermal-hydraulic coupling as part of the ECP project.

smharper commented:
The `set_temperature` and `set_density` methods feel a little odd to me.  Can you explain the thought behind them?  Personally, I think it would make more since to remove those methods and just stick to `cell_set_temperature`, `material_set_density`, and `find`.  Those methods make it crystal clear that OpenMC has the model split into discrete cells with one temperature covering the entire cell.  Even to use the xyz-based method, the TH code needs to be aware of the exact OpenMC geometry so it knows what regions of space it needs to average temperatures over.  And if it knows the exact OpenMC geometry, then why not just set values by cell id's rather than using an xyz intermediate?

smharper commented on src/initialize.F90:
```diff
@@ -59,7 +66,7 @@ subroutine openmc_init(intracomm)
 #ifdef MPI
     ! Setup MPI
-    call initialize_mpi(intracomm)
+    call initialize_mpi(comm)
```
We need to check for a null pointer in case a user called `openmc.lib.init()` without a communicator.

smharper commented:
Also, I think we should consider using the Python `with` feature.  This would mean that API interactions could look something like,
```python
with openmc.lib.open(MPI.COMM_WORLD) as kernel:
    kernel.init()
    for i in range(n_iters):
        do_stuff()
        kernel.run()
        kernel.reset()
    kernel.finalize()
```

The advantage of this is that we can add an `__exit__` method that would guarantee proper clean up of memory and MPI processes in case an exception was called or a user skipped the call to `finalize`

paulromano commented:
Neat idea- adding a context manager. In that case, we'd probably just want `__enter__` to call `init()` and `__exit__` to call `finalize()`. Anyway, I'll think about that some more.

smharper commented:
I'm not positive that we want  `__exit__` to call `finalize()` because we might want to skip the statepoint output.

smharper commented:
On second thought, maybe the proper way to handle that is to add a `statepoint` key to `Settings.output`.

paulromano commented:
@smharper I've addressed all your comments. I'm not seeing any issues with zero entropy or keff NaNs. Can you try again and see if you're still getting them? I removed the `openmc_set_density` and `openmc_set_temperature` functions. They were intended to be an easy way to set properties given a point location, but as you point out combining `find` and `material_set_*` is sufficient.

I also took a first shot at adding a context manager, currently called `lib_context` (really couldn't think of anything better -- let me know if you have ideas).

smharper commented:
I'm still getting the NaN's and zero entropy values.  Any suggestions on how to debug this?  It's probably some MPI communicator thing.  I imagine the first step would be to write a C code that calls this API directly.  FWIW, my build config is GCC-5.3.0, MPICH-3.2, PHDF5-1.8.16

paulromano commented:
Are you sure that mpi4py was built with the same MPI that you used to build OpenMC/PHDF5? To check, you can use:
```Python
import mpi4py
print(mpi4py.get_config())
```

paulromano commented:
Ok, I just set up a new environment with MPICH and I'm also seeing the zeros and NaNs that you're getting, so I'll try to debug as well. Before I was using OpenMPI and didn't see any of these issues, interestingly.

paulromano commented:
It looks like any call to `MPI_REDUCE` with `MPI_IN_PLACE` doesn't work correctly when called from the Python-instantiated OpenMC shared library. I'm really at a loss for why this happens but I may try to boil it down to a simple example and report a bug to mpi4py.

cjosey commented:
I don't really have time to do thorough testing, but one of my concerns was that it might be possible to get an MPI simulation into an inconsistent state:

```
from mpi4py import MPI
import openmc

openmc.lib.init()
if MPI.COMM_WORLD.rank == 0:
    openmc.lib.cell_set_temperature(1, 1000.0)
else:
    openmc.lib.cell_set_temperature(1, 300.0)
openmc.lib.run()
openmc.lib.finalize()
```

I don't see anything that could prevent this that wouldn't harm performance other than "don't do that".  Since this is a low-level API, I expect that just mentioning it in documentation might be enough.

smharper commented:
`lib_context` is a little cryptic.  How about `run_in_memory` or `kernel_in_memory`?

@cjosey, good point, but I agree it would be difficult to guard against rigorously.  This is kind of a power user feature so I think we're probably fine leaving that risk in.

I installed an OpenMPI toolchain and confirmed that it works fine.  I'm happy to merge this PR despite the MPICH issue.  I think we should try to detect MPICH and raise a warning, though.  I'd suggest adding a `try import mpi4py.MPI; catch ImportError` that checks `mpi4py.MPI.get_vendor()` and raises a warning for anything that isn't OpenMPI.  (We'd also have to make it so `openmc.capi` isn't imported by default so that a user will only see that error if they choose to import `openmc.capi`.)

paulromano commented:
Thanks @smharper. I'm still working on some changes/additions 1) to address the MPICH issue, if possible, and 2) for the sake of calling OpenMC from OpenDeplete. I'll also try to address the comments you just made. I'm going to put the 'On Hold' label until this PR is ready for another look.

paulromano commented:
Just pushed some commits that should fix the MPICH problems

paulromano commented:
I've done a fairly thorough refactor of the bindings to the C API. A few of the salient points:

1. The bindings now live in a `openmc.capi` subpackage (it has its own directory)

2. The OpenMCLibrary class is gone. Instead, many of its methods are now just regular functions. Some have been grouped into new classes though. For example, there is now a `MaterialView` class that encapsulates all the material related functionality.

3. There are now a few module variables at the Python level that expose dictionaries at the Fortran level. As one example, `openmc.capi.materials` is a mapping where you index it with the ID of a material and it gives you a `MaterialView` object that allows you to interact with the material as stored in the Fortran derived type. So, if you wanted to get the densities of nuclides inside material 100, it would look like

    ```Python
    m = openmc.capi.materials[100]
    print(m.get_densities())
    ```

    The module variables (e.g. `openmc.capi.materials`) are actually instances of special classes that implement `__getitem__`, `__iter__`, and `__len__` methods which satisfy the mapping protocol. By inheriting from `collections.Mapping`, we get other functionality (mixin methods) such as being able to say `100 in openmc.capi.materials` or `list(openmc.capi.materials.keys())`.

4. Error handling is now very clean -- our C API uses error codes almost exclusively (other than functions that like `openmc_run` that existed pre-C-API. ctypes allows you to set an `errcheck` attribute on a C function pointer that calls an arbitrary function if the C function returns an int, so I wrote an `_error_handler` function that looks at the error code and raises an appropriate exception.

@smharper This is now ready for review once again.

paulromano commented:
@smharper I just addressed your comments. I decided to remove the try/except around importing the shared library. If a user has `import openmc.capi`, it means they probably want to call functions from the C API. Thus, raising an exception if the shared library is not available seems like the right thing to do rather than just giving a warning.

smharper commented on src/api.F90:
```diff
@@ -0,0 +1,890 @@
+    end if
+  end function openmc_tally_set_nuclides
+
+  function to_f_string(c_string) result(f_string)
```
Add a comment block

smharper commented on src/api.F90:
```diff
@@ -0,0 +1,890 @@
+          end if
+        end associate
+
+        ! assign error codes for outside of temperature bounds provided the
```
It looks odd for this multi-sentence comment to start with a lowercase character.

paulromano commented:
@smharper, latest round of comments has been addressed

smharper commented:
Looks, good.  Merging!
